src/arith_uint256.cpp

```diff
@@ -184,6 +184,29 @@ unsigned int base_uint<BITS>::bits() const
     return 0;
 }
 
+template <unsigned int BITS>
+base_uint<BITS> base_uint<BITS>::sqrt() const
+{
+    base_uint<BITS> num = *this;
+    base_uint<BITS> res = 0;
+    base_uint<BITS> bit = base_uint<BITS>(1) << (BITS-2); // The second-to-top bit is set.
+                           // Same as ((unsigned) INT32_MAX + 1) / 2.
+
+    // "bit" starts at the highest power of four <= the argument.
+    while (bit > num)
+        bit >>= 2;
+
+    while (bit != 0) {
+        if (num >= res + bit) {
+            num -= res + bit;
+            res = (res >> 1) + bit;
+        } else
+            res >>= 1;
+        bit >>= 2;
+    }
+    return res;
+}
```

Comment:
Can add some unit test cases for this function?

Reply:
added, see last commit